viewing_party/party.py

```diff
@@ -1,23 +1,270 @@
+            user_data["watched"].append(movie)
+            user_data["watchlist"].remove(movie)
+
+    return user_data
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    '''
+    Returns an int that represents the average rating of the user's
+    watched movies
+    '''
+    watched_rating_list = []
+    total_rating = 0
+
+    user_data_watched = user_data["watched"]
+
+    if user_data_watched == []:
+        return 0.0
+
+    # appends the watched_rating_list with the rating for each movie in the watched list
+    for movie in user_data_watched:
+        watched_rating_list.append(movie['rating'])
+
+    # adds the ratings together and stores them in total_rating
+    for rating in watched_rating_list:
+        total_rating += rating
+
+    avg_rating = total_rating / len(watched_rating_list)
+
+    return avg_rating
+
+def get_most_watched_genre(user_data):
+    '''
+    Finds the most watched genre from the user's watched list 
+    and returns a string that represents that genre
+    '''
+    watched_genre_count = {}
+    most_watched_genre_total = 0
+    most_watched_genre = ""
+
+    user_data_watched = user_data["watched"]
+
+    if user_data_watched == []:
+        return None
+
+    for movie in user_data_watched:
+        if movie["genre"] in watched_genre_count:
+            watched_genre_count[movie["genre"]] += 1
+        else:
+            watched_genre_count[movie["genre"]] = 1
+
+    for genre in watched_genre_count:
+        if watched_genre_count[genre] > most_watched_genre_total:
+            most_watched_genre_total = watched_genre_count[genre]
+            most_watched_genre = genre
+
+    return most_watched_genre
+
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    '''
+    collects uniqe movies watched by user compared to friends and
+    stores it in a list, then returns that list
+    '''
+    unique_watched_list = []
+
+    friend_watched_set = set(get_friends_watched(user_data))
+    user_watched_set = set(get_users_watched(user_data))
+
+    unique_set = user_watched_set - friend_watched_set
+
+    for i in range(len(user_data["watched"])):
+        if user_data["watched"][i]['title'] in unique_set:
+            unique_watched_list.append(user_data["watched"][i])   
+
+    return unique_watched_list
+    
+    
+    
+def get_friends_watched(user_data):
+    '''
+    Helper Function - Stores the titles of the friends 'watched' list from 
+    user_data in a list, and returns that list
+    '''   
+    friends_list_of_dict = user_data["friends"]
+    friends_watched_list = []
+
+    for i in range(len(friends_list_of_dict)):
+        for watched in friends_list_of_dict[i]:
+            for movie in friends_list_of_dict[i][watched]:

```

Comment:
There is a small potential bug here.  This second for loop is looping over all of the keys in a friends dictionary, and it works here because in our examples friends only have one key, `watched`.  Friends could theoretically have a "watchlist" or "subscriptions" or other data as part of the dictionary, and in those cases, this loop would pull in extra unwanted data.  However, even if we define a friend dictionary to only have the one key, this middle loop is not necessary.  The key-value pair for the 'watched' key can be accessed directly using square brackets:
```suggestion
    for i in range(len(friends_list_of_dict)):
        watched = friends_list_of_dict[i]["watched"]
        for movie in watched:

```